sccp/parafac2.py

```diff
@@ -1,8 +1,29 @@
+from typing import Sequence
 import numpy as np
 import anndata
 from pacmap import PaCMAP
 from parafac2 import parafac2_nd
 from .imports import import_citeseq, import_lupus, import_thomson
+from tensorly.parafac2_tensor import parafac2_to_slice
+
+
+def cwSNR(
+    matrices: Sequence,
+    weights: np.ndarray,
+    factors: list[np.ndarray],
+    projections: list[np.ndarray],
+):
+    """Calculate the columnwise signal-to-noise ratio for each dataset."""
+    SNR = np.empty((len(matrices), len(weights)), dtype=float)
+
+    for i, mat in enumerate(matrices):
+        xx = parafac2_to_slice((weights, factors, projections), i, validate=False)
+
+        E_k = np.linalg.norm(mat - xx)
+
+        SNR[i, :] = factors[0][i, :] * weights / E_k
```

Comment:
@aarmey Doesn't the numerator need to be squared as well based on the equation?

Reply:
Good catch! I'll fix this on main.